### ISSUE TITLE
Building libcxxabi throws “cannot find -lcxx-headers” error

### ISSUE THREAD

**Issue by jonathonmckay** [llvm/llvm-project#116088]
Apologies in advance if this isn't the right place to raise this.
I maintain a small HPC cluster. As a result, I often have to install things in non standard ways.
I already have LLVM and Clang 16.0.6 compiled and installed. While attempting to build a 3rd party program, it became clear that I also needed libc++ and libc++abi.

I found instructions on installing libc++ and libc++abi on CentOS7 from Stack Overflow [here](https://stackoverflow.com/questions/25840088/how-to-build-libcxx-and-libcxxabi-by-clang-on-centos-7/25840107#25840107). These are essentially the steps I followed, making path and other settings changes as needed. I downloaded the .src.tar.xz files for libcxx and libcxxabi from https://github.com/llvm/llvm-project/releases/tag/llvmorg-16.0.6

Step 5 involves building libcxx without libcxxabi. After some minor problems, I got the following cmake command to do what I want:
`cmake -DCMAKE_INSTALL_PREFIX=/minerva/opt/llvm-16.0.6/ -DCMAKE_BUILD_TYPE=Release -DLIBCXX_INCLUDE_BENCHMARKS=OFF -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DLIBCXX_CXX_ABI=none ..`

Once that worked and was built and installed with `make install`, I then moved on to Step 6, which involves building libcxxabi with the libcxx I just built. After a few more minor problems, I got the following cmake command:
`cmake -DCMAKE_INSTALL_PREFIX=/minerva/opt/llvm-16.0.6/ -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DLIBCXXABI_LIBCXX_INCLUDES=../../libcxx-16.0.6.src/include/ -DLIBCXXABI_LIBCXX_PATH=../../libcxx-16.0.6.src/  ..`
This finished successfully.

Then when trying to actually compile with `make`, while it is trying to link libc++abi.so, I get the following error:
`/usr/bin/ld: cannot find -lcxx-headers`

From what I can tell, this is supposed to be part of libcxx, but I cannot find any "libcxx-headers" in either the source code or in the installed location.
I found that libcxx has make targets generate-cxx-headers, but I verified those have been done and finished without errors.

I expect I'm missing something fundamental, and have likely done something somewhat daft due to the way we install things on our HPC system.

If anyone has any insight, it would be greatly appreciated (There has already been two days of searching and trying things). If any additional information is required, please don't hesitate to ask.

**Comment by ldionne** [comment]
Can you try using a recent libc++ (LLVM 19 or even better, `main`) with the instructions at https://libcxx.llvm.org/VendorDocumentation.html#the-default-build instead? LLVM 16 is fairly old and I don't know what might be going wrong, the build system evolves a lot at each release. If you encounter issues with the latest release or `main`, please ping this issue and we should be able to help.

Closing in the meantime, please reopen if you have additional information / want us to take a look.

**Comment by reactive-firewall** [comment]
## Regarding insights:

@jonathonmckay

> If anyone has any insight, it would be greatly appreciated (There has already been two days of searching and trying things). If any additional information is required, please don't hesitate to ask.

> I don't know what might be going wrong ...
> ... If you encounter issues with the latest release or `main` ...

I am able to reproduce this behavior on latest LLVM 21.1.5 release; and I'm not sure where else to document my insights, so I'm putting them here. @ldionne, Consider this FYI, and No need to reopen this GHI.

Insights on steps to reproduce this issue:
1 build/provide a supported libc, compiler runtime (e.g. compiler-rt (namely need the builtins for step 2))
2 build/provide an unwinder (I see this more with libunwind, as libgcc is coupled tighter and difficult to seperate anyway, but this can be reproduced with libcxxrt's itainium unwinder (did not test the arm variant much))
3 build just libcxx target from llvm, either via `LIBCXX_CXX_ABI=none` or `LIBCXX_CXX_ABI=libcxxrt` (and with the stuff from steps 1 & 2, (issue much more common on cross-builds, as things like the c++ tests often won't let you get this far otherwise))
4 try to build libcxxabi against the new libcxx artifacts; see error occur.

**The short**: AFAIK, is the bootstrap build does not _officially support_ (as in documentation does not indicate it as a supported use-case) individually building libcxxabi and libcxx separately. :shrug:

---

## Additional workaround insights

TL;DR - you will need to bootstrap the two components first and then re-build them (semi-round-trip bootstrap style) via the runtimes target to get around this issue. The root of the issue is arising from how `cmake` resolves the mutual dependencies configured for the libcxx and libcxxabi targets. When following the documented typical build, the runtimes target is used instead of directly targeting either libcxx or libcxxabi (also optionally libunwind too, btw). This works fine for the documented and 'supported' build use-case.

TL;DR - You can bootstrap the libcxx (via `LIBCXX_CXX_ABI=none` or `LIBCXX_CXX_ABI=libcxxrt`), and then point your libcxxabi at the bootstrapped libc++.{a,so} (and typically the libc++experimental.a) to build the static (e.g. `libc++ability.a`) OR try my workaround for tricking `cmake` into building the shared lib.

TL;DR - The way I go around this was to use the obscure `DT_AXILLARY` Extensible Linked Format (ELF) dynamic header to provide a "missing" `libcxx-headers.so` for cmake to fall-back on. The key is that when linking (e.g., with LLVM LLD) my _empty_ `libcxx-headers.so` I do so with:
```console
# ${HOST_CC} ... -fuse-ld=lld ... -o libcxx-headers.so \
      -Xlinker --auxiliary=libc++.so \
      -Xlinker --auxiliary=libcxxrt.so \
      -Xlinker --auxiliary=libc.so \
      ... etc.
```

Note: The `-Xlinker --auxiliary=libcxxrt.so` is only needed for the `LIBCXX_CXX_ABI=libcxxrt` case, and obviously requires a shared build of libcxxrt (in step 2 of reproducing the issue).

Once I have the bootstrapped `libc++abi.so` linked to my `libcxx-headers.so` I just rebuild via the runtimes target (in a fresh build environment to avoid cmake pulling in other stuff) specifically enabling `LLVM_ENABLE_RUNTIMES="libcxx;libcxxabi"` as if it were a typical build, and providing the bootstraped `libc++*` libs to complete the round-trip bootstrap build.

---

If anyone else stumbles upon this GHI, hopefully, they will find these insights helpful.
